core/eolearn/core/eodata_io.py

```diff
@@ -221,18 +229,30 @@ def _to_lowercase(ftype, fname, *_):
     return ftype, fname if fname is ... else fname.lower()
 
 
+def _get_file_format(feature_type: FeatureType) -> FileFormat:
+    """ For each feature type it decides to which file format it will be serialized
+    """
+    if feature_type is FeatureType.BBOX:
+        return FileFormat.GEOJSON
+    if feature_type.is_raster():
+        return FileFormat.NPY
+    if feature_type.is_vector():
+        return FileFormat.GPKG
+    return FileFormat.JSON
+
+
 class FeatureIO:
     """ A class that handles the saving and loading process of a single feature at a given location
     """
-    def __init__(self, filesystem, path):
+    def __init__(self, filesystem: fs.base.FS, path: str, feature_type: FeatureType):
```

Comment:
As above: I'd prefer signature `feature_type: FeatureType, filesystem: fs.base.FS, path: str`. Even better would be that `FeatureIO` doesn't have to know its feature type, but that could only be achieved if `FeatureIO` would be a part fo feature itself. Unfortunately, that cannot be achieved without a substantial refactoring (and code breaking changes).

Reply:
I agree about the order (or at least I don't have a strong preference).

I wanted to avoid adding `feature_type` parameter here but it was necessary. As you can see it is required for:

- additional serialization/deserialization steps for certain feature types,
- nicer error messages.

Reply:
> I wanted to avoid adding feature_type parameter here but it was necessary.

Indeed; `FeatureIO`, `FeatureType` and `FileFormat` are very intertwined. I believe there is better architectural construct to tie them together where adding parameters won't be necessary.